core/src/main/scala/org/apache/spark/TaskContextImpl.scala

```diff
@@ -275,6 +275,8 @@ private[spark] class TaskContextImpl(
   @GuardedBy("this")
   override def isCompleted(): Boolean = synchronized(completed)
 
+  override def isFailed(): Boolean = synchronized(failureCauseOpt.isDefined)
+
   override def isInterrupted(): Boolean = reasonIfKilled.isDefined
```

Comment:
I'm trying to reason about the relationship between these 3 flags:
- `isCompleted` is false only when the task is interrupted? Or when the task is still running?
- `isFailed` is true only if the task completes and fails?
- `isInterrupted` can be true before the task completes?

Reply:
`isCompleted` includes all the state `success`, `failure` and `cancellation`. If isCompleted is true, then the task won't be in running any more.